src/language/branches.md

```diff
@@ -0,0 +1,295 @@
+
+  nowClosed:
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+### Switch expressions
+
+_Switch expressions_ write a result value determined by a switch
+directly into a variable or return statement. 
+
+The syntax rules for a switch expression are:
+- The expression body is a series of case patterns.
+- Cases _do not_ use the `case` keyword.
+- Case patterns are followed by `=>`, and then a case body. 
+- Case bodies terminate with a comma `,`.
+- The switch expression body terminates with a semicolon `;`.
+- The wildcard pattern `_` catches remaining values instead of `default`.
+
+This allows you to rewrite a 
+_statement_ like this:
+
+```dart
+switch (charCode) {
+  case slash when nextCharCode == slash:   // Constant pattern with guard
+    skipComment();
+  case slash || star || plus || minus:     // Logical-or pattern
+    operator(charCode);
+  case >= digit0 && <= digit9:             // Relational and logical-and patterns
+    number();
+  default:
+    invalid();
+}
+```
+
+Into an _expression_, like this:
+
+```dart
+var token = switch (charCode) {
+  slash when nextCharCode == slash => skipComment(),
+  slash || star || plus || minus   => operator(charCode),
+  >= digit0 && <= digit9           => number(),
+  _                                => throw invalid()
+};
+```
+
+### Exhaustiveness checking
+
+Exhaustiveness checking is a feature of switches that raises a compile-time error
+if a switch is not exhaustively covering all possible values that could flow through it.
+
+The compiler can only know if a set of cases is
+exhaustive if it is soundly aware of all possible enumerations of that type. So,
+exhaustiveness checking works only when switching on types that are fully enumerable:
+
+- [Boolean][boolean] types, because there are only two possible values, `true` and `false`.
+- [Enum][enum] types, because they are always a fixed number of constant values.
+- [Sealed][sealed] types, because they are classes with a closed hierarcy of subtypes in the same library.
+
+```dart
+bool? b = false;
+// Non-exhaustive switch on bool?, missing case to match null possiblity:
+switch (b) {
+  case true: print('yes');
+  case false: print('no');
+}
+```
+
+Use the [`sealed` modifier][sealed] on a class to enable exhaustiveness checking when 
+switching over subtypes of that class:
+
+```dart
+sealed class Shape {
+  double calculateArea();
+}
+
+class Square extends Shape { ... }
+class Circle extends Shape { ... }
+
+// ...
+
+double calculateArea(Shape shape) =>
+  switch (shape) {
+    Square(length: var l) => l * l,
+    Circle(radius: var r) => math.pi * r * r
+  };
+```
+
+If a new subclass of `Shape` is ever added, exhaustiveness 
+checking will warn you about the missing subtype, because your switch is incomplete.
+This allows you to use Dart in a somewhat 
+[functional algrbraic datatype style](https://en.wikipedia.org/wiki/Algebraic_data_type). 
+
```

Comment:
Then we'd need to say that switch expressions are always required to be exhaustive, and that this is possible even with types that aren't enumerable if there is a case that covers all possible values (such as a default case or a case like `var x` which will match every value of the scrutinee and give `x` the same declared type as the static type of the scrutinee expression).